Automattic/wp-calypso
# Framework: Preload moment locale data when selected language different than `en`

This pull request seeks to retake work done by @stephanethomas and @gziolo in https://github.com/Automattic/wp-calypso/pull/12491

## Problem

[`moment`](https://github.com/moment/moment) module is used for localizing dates and times. It comes with lots of `.js` files, one for each supported language. All language files are bundled by Webpack into the `vendor` chunk.

![image](https://user-images.githubusercontent.com/4490611/31306553-d5a616b8-ab52-11e7-86ca-5e0d7ac0662e.png)

## Solving

* Setup Wepbpack to ignore bundling of `moment` locales, as explained in [`IgnorePlugin`'s documentation](https://webpack.js.org/plugins/ignore-plugin/#ignore-moment-locales).

* Change the server code to load the correct locale file before the chunk is loaded, as explained [here](https://github.com/Automattic/wp-calypso/pull/12491#issuecomment-303736837).

* Setup Webpack to use a custom loader to patch `i18n-calypso` module. `i18n-calyspo` code is changed so that if locale is different than `en`, use `require` with a [`bundle-loader`](https://www.npmjs.com/package/bundle-loader) query to request the locale bundle. That will use a callback to set the correct locale when the locale chunk is loaded.

## Testing

* Run git checkout try/lazyload-moment and start your server, or open a live branch
* Log in to an English account
* Check that the vendor bundle is now ~2.3M (instead of ~2.5M) in the network panel of your browser
* Check that no moment-locale bundle was loaded in this panel
* Navigate to the Account Settings page
* Notice the date in English just below the Username field (e.g. Joined April 2016)
* Select a different language in the Interface Language select box
* Hit the Save Account Settings button, which will reload the page
* Check that a moment-locale was loaded this time in the network panel
* Check that the date below the Username field is now in the language chosen

Note that in logged out routes, the language will always be English due to [this](https://github.com/Automattic/wp-calypso/issues/14508) bug.
